File: machine/ts/machine/devices/irq/controller.ts
import { IO_IRQ_ACK, IO_IRQ_FLAGS, IO_IRQ_MASK } from '../../../spec/bmsx/io';
import type { Value } from '../../cpu/value';
import { Memory } from '../../memory/memory';
import type { IrqControllerState } from './save_state';

export class IrqController {
	private pendingFlags = 0;
	private mask = 0;
	private userPendingFlags = 0;
	private userMask = 0;
	private supervisorContextActive = false;

	public constructor(private readonly memory: Memory) {
		this.memory.mapIoRead(IO_IRQ_FLAGS, this, IrqController.onFlagsReadThunk);
		this.memory.mapIoWrite(IO_IRQ_ACK, this, IrqController.onAckWriteThunk);
		this.memory.mapIoRead(IO_IRQ_MASK, this, IrqController.onMaskReadThunk);
		this.memory.mapIoWrite(IO_IRQ_MASK, this, IrqController.onMaskWriteThunk);
	}

	public reset(): void {
		this.pendingFlags = 0;
		this.mask = 0;
		this.userPendingFlags = 0;
		this.userMask = 0;
		this.supervisorContextActive = false;
		this.memory.writeIoValue(IO_IRQ_ACK, 0);
		this.memory.writeIoValue(IO_IRQ_MASK, 0);
	}

	public postLoad(): void {
		const clearAck = 0;
		this.memory.writeIoValue(IO_IRQ_ACK, clearAck);
		this.memory.writeIoValue(IO_IRQ_MASK, this.mask);
	}

	public captureState(): IrqControllerState {
		return {
			mask: this.mask,
			pendingFlags: this.pendingFlags,
			userMask: this.userMask,
			userPendingFlags: this.userPendingFlags,
			supervisorContextActive: this.supervisorContextActive,
		};
	}

	public restoreState(state: IrqControllerState): void {
		this.mask = state.mask >>> 0;
		this.pendingFlags = state.pendingFlags >>> 0;
		this.userMask = state.userMask >>> 0;
		this.userPendingFlags = state.userPendingFlags >>> 0;
		this.supervisorContextActive = state.supervisorContextActive;
		this.postLoad();
	}

	public enterSupervisorContext(): void {
		this.userPendingFlags = this.pendingFlags;
		this.userMask = this.mask;
		this.supervisorContextActive = true;
		this.pendingFlags = 0;
		this.mask = 0;
		this.postLoad();
	}

	public enterSupervisorFaultContext(): void {
		this.userPendingFlags = 0;
		this.userMask = 0;
		this.supervisorContextActive = true;
		this.pendingFlags = 0;
		this.mask = 0;
		this.postLoad();
	}

	public leaveSupervisorContext(): void {
		this.pendingFlags = this.userPendingFlags;
		this.mask = this.userMask;
		this.userPendingFlags = 0;
		this.userMask = 0;
		this.supervisorContextActive = false;
		this.postLoad();
	}

	public hasAssertedMaskableInterruptLine(): boolean {
		return (this.pendingFlags & this.mask) !== 0;
	}

	public raise(mask: number): void {
		const next = (this.pendingFlags | (mask >>> 0)) >>> 0;
		if (next !== this.pendingFlags) {
			this.pendingFlags = next;
		}
	}

	public raiseUser(mask: number): void {
		if (!this.supervisorContextActive) {
			this.raise(mask);
			return;
		}
		this.userPendingFlags = (this.userPendingFlags | (mask >>> 0)) >>> 0;
	}

	public acknowledge(mask: number): void {
		const ack = mask >>> 0;
		if (ack !== 0) {
			const next = (this.pendingFlags & ~ack) >>> 0;
			if (next !== this.pendingFlags) {
				this.pendingFlags = next;
			}
		}
		this.memory.writeIoValue(IO_IRQ_ACK, 0);
	}

	private static onFlagsReadThunk(context: IrqController, addr: number): Value {
		void addr;
		return context.pendingFlags;
	}

	private static onAckWriteThunk(context: IrqController, addr: number, value: Value): void {
		void addr;
		const ack = (value as number) >>> 0;
		if (ack !== 0) {
			const next = (context.pendingFlags & ~ack) >>> 0;
			if (next !== context.pendingFlags) {
				context.pendingFlags = next;
			}
		}
		context.memory.writeIoValue(IO_IRQ_ACK, 0);
	}

	private static onMaskReadThunk(context: IrqController, addr: number): Value {
		void addr;
		return context.mask;
	}

	private static onMaskWriteThunk(context: IrqController, addr: number, value: Value): void {
		void addr;
		context.mask = (value as number) >>> 0;
	}
}

File: machine/cpp/machine/devices/irq/controller.cpp
#include "machine/devices/irq/controller.h"
#include "spec/bmsx/io.h"

namespace bmsx {

IrqController::IrqController(Memory& memory)
	: m_memory(memory) {
	m_memory.mapIoRead(IO_IRQ_FLAGS, this, &IrqController::onFlagsReadThunk);
	m_memory.mapIoWrite(IO_IRQ_ACK, this, &IrqController::onAckWriteThunk);
	m_memory.mapIoRead(IO_IRQ_MASK, this, &IrqController::onMaskReadThunk);
	m_memory.mapIoWrite(IO_IRQ_MASK, this, &IrqController::onMaskWriteThunk);
}

void IrqController::reset() {
	m_pendingFlags = 0;
	m_mask = 0;
	m_userPendingFlags = 0;
	m_userMask = 0;
	m_supervisorContextActive = false;
	m_memory.writeIoU32(IO_IRQ_ACK, 0u);
	m_memory.writeIoU32(IO_IRQ_MASK, 0u);
}

void IrqController::postLoad() {
	m_memory.writeIoU32(IO_IRQ_ACK, 0u);
	m_memory.writeIoU32(IO_IRQ_MASK, m_mask);
}

void IrqController::raise(uint32_t mask) {
	const uint32_t next = m_pendingFlags | mask;
	if (next != m_pendingFlags) {
		m_pendingFlags = next;
	}
}

void IrqController::raiseUser(uint32_t mask) {
	if (!m_supervisorContextActive) {
		raise(mask);
		return;
	}
	m_userPendingFlags |= mask;
}

void IrqController::enterSupervisorContext() {
	m_userPendingFlags = m_pendingFlags;
	m_userMask = m_mask;
	m_supervisorContextActive = true;
	m_pendingFlags = 0u;
	m_mask = 0u;
	postLoad();
}

void IrqController::leaveSupervisorContext() {
	m_pendingFlags = m_userPendingFlags;
	m_mask = m_userMask;
	m_userPendingFlags = 0u;
	m_userMask = 0u;
	m_supervisorContextActive = false;
	postLoad();
}

void IrqController::acknowledge(uint32_t mask) {
	if (mask != 0u) {
		const uint32_t next = m_pendingFlags & ~mask;
		if (next != m_pendingFlags) {
			m_pendingFlags = next;
		}
	}
	m_memory.writeIoU32(IO_IRQ_ACK, 0u);
}

u32 IrqController::onFlagsReadThunk(void* context, [[maybe_unused]] uint32_t addr, MappedBusSignals) {
	return static_cast<IrqController*>(context)->m_pendingFlags;
}

void IrqController::onAckWriteThunk(void* context, [[maybe_unused]] uint32_t addr, u32 value, MappedBusSignals) {
	auto& irq = *static_cast<IrqController*>(context);
	if (value != 0u) {
		const uint32_t next = irq.m_pendingFlags & ~value;
		if (next != irq.m_pendingFlags) {
			irq.m_pendingFlags = next;
		}
	}
	irq.m_memory.writeIoU32(IO_IRQ_ACK, 0u);
}

u32 IrqController::onMaskReadThunk(void* context, [[maybe_unused]] uint32_t addr, MappedBusSignals) {
	return static_cast<IrqController*>(context)->m_mask;
}

void IrqController::onMaskWriteThunk(void* context, [[maybe_unused]] uint32_t addr, u32 value, MappedBusSignals) {
	static_cast<IrqController*>(context)->m_mask = value;
}

} // namespace bmsx
